File: android/src/main/java/com/reactnativechimesdk/ChimeSdkModule.java
package com.reactnativechimesdk;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.hardware.camera2.CameraManager;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;

import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.contentshare.ContentShareObserver;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.contentshare.ContentShareStatus;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.capture.CameraCaptureSource;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.capture.DefaultCameraCaptureSource;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.capture.DefaultSurfaceTextureCaptureSourceFactory;
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.capture.VideoCaptureFormat;
import com.amazonaws.services.chime.sdk.meetings.device.DeviceChangeObserver;
import com.amazonaws.services.chime.sdk.meetings.device.MediaDevice;
import com.amazonaws.services.chime.sdk.meetings.device.MediaDeviceType;
import com.amazonaws.services.chime.sdk.meetings.session.DefaultMeetingSession;
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSession;
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionConfiguration;
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus;
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatusCode;
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger;
import com.amazonaws.services.chime.sdk.meetings.utils.logger.LogLevel;
import com.annimon.stream.Stream;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.google.gson.Gson;
import com.reactnativechimesdk.data.RosterAttendee;
import com.reactnativechimesdk.data.VideoCollectionTile;
import com.reactnativechimesdk.itf.SimpleAudioVideoObserver;
import com.reactnativechimesdk.itf.SimpleRealTimeObserver;
import com.reactnativechimesdk.itf.SimpleVideoTileObserver;
import com.reactnativechimesdk.response.JoinMeetingResponse;
import com.reactnativechimesdk.utils.Util;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import static android.Manifest.permission.CAMERA;
import static android.Manifest.permission.MODIFY_AUDIO_SETTINGS;
import static android.Manifest.permission.RECORD_AUDIO;
import static androidx.core.content.ContextCompat.checkSelfPermission;
import static com.reactnativechimesdk.EventEmitter.KEY_AUDIO_STATUS;
import static com.reactnativechimesdk.EventEmitter.KEY_USER_ID;
import static com.reactnativechimesdk.EventEmitter.KEY_USER_NAME;
import static com.reactnativechimesdk.EventEmitter.KEY_VIDEO_STATUS;
import static com.reactnativechimesdk.EventEmitter.MEETING_ACTIVE_SHARE;
import static com.reactnativechimesdk.EventEmitter.MEETING_AUDIO_STATUS_CHANGE;
import static com.reactnativechimesdk.EventEmitter.MEETING_USER_JOIN;
import static com.reactnativechimesdk.EventEmitter.MEETING_USER_LEFT;
import static com.reactnativechimesdk.EventEmitter.MEETING_VIDEO_STATUS_CHANGE;
import static com.reactnativechimesdk.EventEmitter.SHARE_STATUS_START;
import static com.reactnativechimesdk.EventEmitter.SHARE_STATUS_STOP;
import static com.reactnativechimesdk.EventEmitter.sendMeetingStateEvent;
import static com.reactnativechimesdk.EventEmitter.sendMeetingUserEvent;
import static com.reactnativechimesdk.EventEmitter.sendMeetingUserShareEvent;
import static com.reactnativechimesdk.MeetingModel.meetingModel;
import static com.reactnativechimesdk.response.Api.createSession;
import static com.reactnativechimesdk.response.Api.requestCreateSession;
import static com.reactnativechimesdk.utils.Util.getAttendeeName;

public class ChimeSdkModule extends ReactContextBaseJavaModule
  implements LifecycleEventListener, SimpleRealTimeObserver, SimpleAudioVideoObserver, SimpleVideoTileObserver, DeviceChangeObserver, ContentShareObserver {

  private static final String TAG = "ChimeSdkModule";
  private static final ConsoleLogger logger = new ConsoleLogger(LogLevel.INFO);

  private static final int MAX_VIDEO_FORMAT_HEIGHT = 240;
  private static final int MAX_VIDEO_FORMAT_FPS = 15;

  private final int WEBRTC_PERMISSION_REQUEST_CODE = 1030;
  private final List<String> WEBRTC_PERM = Arrays.asList(MODIFY_AUDIO_SETTINGS, RECORD_AUDIO, CAMERA);
  private final IntentFilter intentFilter = new IntentFilter() {
    {
      addAction("onConfigurationChanged");
      addAction("onRequestPermissionsResult");
    }
  };

  private final BroadcastReceiver moduleConfigReceiver = new BroadcastReceiver() {
    @Override
    public void onReceive(Context context, Intent intent) {
      if ("onRequestPermissionsResult".equals(intent.getAction())) {
        String[] permissions = intent.getStringArrayExtra("permissions");
        int[] grantResults = intent.getIntArrayExtra("grantResults");
        int requestCode = intent.getIntExtra("requestCode", -1);
        if (permissions == null || grantResults == null || requestCode != WEBRTC_PERMISSION_REQUEST_CODE) {
          return;
        }
        joinMeeting();
      }
    }
  };

  public ChimeSdkModule(@Nullable ReactApplicationContext reactContext) {
    super(reactContext);
    assert reactContext != null;
    reactContext.addLifecycleEventListener(this);
  }

  private final Gson gson = new Gson();
  private JoinMeetingResponse joinMeetingResponse;

  @NonNull
  @Override
  public String getName() {
    return "ChimeSdk";
  }

  @ReactMethod
  public void showToast(String msg) {
    Toast.makeText(getReactApplicationContext(), msg, Toast.LENGTH_LONG).show();
  }

  @ReactMethod
  public void initSdk() {
    // Ignore
  }

  @ReactMethod
  public void joinMeeting(ReadableMap map) {
    try {
      JSONObject object = Util.convertMapToJson(map);
      joinMeetingResponse = gson.fromJson(object.toString(), JoinMeetingResponse.class);

      // For example only
      if (joinMeetingResponse.joinInfo == null) {
        String url = map.getString("meetingUrl");
        String id = map.getString("meetingId");
        String name = map.getString("attendeeName");
        String rs = requestCreateSession(url, id, name);
        joinMeetingResponse = gson.fromJson(rs, JoinMeetingResponse.class);
      }

      if (joinMeetingResponse != null && joinMeetingResponse.joinInfo != null) {
        meetingModel().setLocalId(joinMeetingResponse.joinInfo.attendeeResponse.attendee.getAttendeeId());
      } else {
        throw new IllegalArgumentException("Error parsing meeting response");
      }
    } catch (IOException | JSONException | IllegalArgumentException e) {
      Log.e(TAG, "Failed to create meeting response", e);
      return;
    }

    if (hasPermissionsAlready()) {
      joinMeeting();
    } else {
      ActivityCompat.requestPermissions(
        Objects.requireNonNull(getCurrentActivity()),
        Stream.of(WEBRTC_PERM).toArray(String[]::new),
        WEBRTC_PERMISSION_REQUEST_CODE);
    }
  }

  private boolean hasPermissionsAlready() {
    return Stream.of(WEBRTC_PERM)
      .filter(p -> checkSelfPermission(getReactApplicationContext(), p) != PackageManager.PERMISSION_GRANTED)
      .toList()
      .isEmpty();
  }

  private void joinMeeting() {
    if (joinMeetingResponse != null) {
      MeetingSessionConfiguration config = createSession(joinMeetingResponse);
      if (config == null) {
        showToast("Failed to join meeting");
        return;
      }
      MeetingSession meetingSession = new DefaultMeetingSession(config, logger, getReactApplicationContext(), meetingModel().eglCoreFactory);
      meetingModel().setMeetingSession(meetingSession);
      meetingModel().setCameraCaptureSource(initCameraCaptureSource());
      meetingModel().getAudioVideo().addRealtimeObserver(this);
      meetingModel().getAudioVideo().addAudioVideoObserver(this);
      meetingModel().getAudioVideo().addVideoTileObserver(this);
      meetingModel().getAudioVideo().addDeviceChangeObserver(this);
      meetingModel().getAudioVideo().addContentShareObserver(this);
      meetingModel().startMeeting();
    } else {
      showToast("Failed to join meeting");
    }
  }

  private CameraCaptureSource initCameraCaptureSource() {
    DefaultSurfaceTextureCaptureSourceFactory surface = new DefaultSurfaceTextureCaptureSourceFactory(logger, meetingModel().eglCoreFactory);
    CameraCaptureSource cameraCaptureSource = new DefaultCameraCaptureSource(getReactApplicationContext(), logger, surface);
    try {
      CameraManager cameraManager = (CameraManager) getReactApplicationContext().getSystemService(Context.CAMERA_SERVICE);
      List<MediaDevice> mediaDevices = MediaDevice.Companion.listVideoDevices(cameraManager);
      MediaDevice mediaDevice = Stream.of(mediaDevices)
        .filter(it -> it.getType() == MediaDeviceType.VIDEO_FRONT_CAMERA)
        .findFirstOrElse(Stream.of(mediaDevices).findFirst().orElseThrow());
      cameraCaptureSource.setDevice(mediaDevice);
      assert cameraCaptureSource.getDevice() != null;
      List<VideoCaptureFormat> formats = MediaDevice.Companion.listSupportedVideoCaptureFormats(cameraManager, cameraCaptureSource.getDevice());
      Stream.of(formats)
        .filter(it -> it.getHeight() <= MAX_VIDEO_FORMAT_HEIGHT)
        .findFirst()
        .ifPresent(it -> cameraCaptureSource.setFormat(new VideoCaptureFormat(it.getWidth(), it.getHeight(), MAX_VIDEO_FORMAT_FPS)));
    } catch (Exception e) {
      Log.e(TAG, "Media devices not found");
    }
    return cameraCaptureSource;
  }

  @ReactMethod
  public void leaveCurrentMeeting() {
    sendMeetingStateEvent(getReactApplicationContext(), "meeting_end");
    meetingModel().endMeeting();
  }

  @ReactMethod
  public void getParticipants(Callback callback) {
    WritableArray array = new WritableNativeArray();
    Stream.of(meetingModel().getCurrentRoster().values())
      .forEach(it -> {
        WritableMap map = new WritableNativeMap();
        map.putString(KEY_USER_ID, it.getAttendeeId());
        map.putString(KEY_USER_NAME, it.getAttendeeName());
        map.putBoolean(KEY_AUDIO_STATUS, it.isMuted());
        map.putBoolean(KEY_VIDEO_STATUS, meetingModel().isCameraAttendeeOn(it.getAttendeeId()));
        array.pushMap(map);
      });
    callback.invoke(array);
  }

  @ReactMethod
  public void getUserInfo(String attendeeId, Callback callback) {
    Stream.of(meetingModel().getCurrentRoster().values())
      .filter(it -> it.getAttendeeId().equals(attendeeId))
      .findFirst()
      .ifPresent(attendee -> {
        WritableMap map = new WritableNativeMap();
        map.putString(KEY_USER_ID, attendee.getAttendeeId());
        map.putString(KEY_USER_NAME, attendee.getAttendeeName());
        map.putBoolean(KEY_AUDIO_STATUS, attendee.isMuted());
        map.putBoolean(KEY_VIDEO_STATUS, meetingModel().isCameraAttendeeOn(attendee.getAttendeeId()));
        callback.invoke(map);
      });
  }

  @ReactMethod
  public void onMyAudio() {
    meetingModel().onOffAudio(true);
  }

  @ReactMethod
  public void offMyAudio() {
    meetingModel().onOffAudio(false);
  }

  @ReactMethod
  public void onOffMyVideo() {
    meetingModel().onOffVideo();
  }

  @ReactMethod
  public void switchMyCamera() {
    meetingModel().switchCamera();
  }

  @Override
  public void onAudioSessionStarted(boolean reconnecting) {
    meetingModel().initMediaDevice();
  }

  @Override
  public void onAudioSessionStopped(@NonNull @NotNull MeetingSessionStatus sessionStatus) {
    if (sessionStatus.getStatusCode() != MeetingSessionStatusCode.OK) {
      meetingModel().endMeeting();
    }
  }

  @Override
  public void onAudioDeviceChanged(@NotNull List<MediaDevice> list) {
    meetingModel().selectAudioDevice(list);
  }

  @Override
  public void onAttendeesJoined(@NonNull AttendeeInfo[] attendeeInfos) {
    Stream.of(attendeeInfos).forEach(it -> {
      meetingModel().getCurrentRoster().put(
        it.getAttendeeId(),
        new RosterAttendee(it.getAttendeeId(), getAttendeeName(it.getExternalUserId()))
      );
      RosterAttendee newAttendee = meetingModel().getAttendee(it.getAttendeeId());
      if (meetingModel().isLocal(newAttendee.getAttendeeId())) {
        Log.d(TAG, "local attendee joined: " + it.getAttendeeId());
        sendMeetingStateEvent(getReactApplicationContext(), "meeting_ready");
      } else {
        Log.d(TAG, "remote attendee joined: " + it.getAttendeeId());
        boolean isCameraOn = meetingModel().isCameraAttendeeOn(it.getAttendeeId());
        sendMeetingUserEvent(getReactApplicationContext(), MEETING_USER_JOIN, newAttendee, isCameraOn);
      }
    });
  }

  @Override
  public void onAttendeesLeft(@NonNull AttendeeInfo[] attendeeInfos) {
    Stream.of(attendeeInfos).forEach(it -> {
      RosterAttendee removal = meetingModel().getCurrentRoster().remove(it.getAttendeeId());
      if (removal != null) {
        boolean isCameraOn = meetingModel().isCameraAttendeeOn(it.getAttendeeId());
        sendMeetingUserEvent(ChimeSdkModule.this.getReactApplicationContext(), MEETING_USER_LEFT, removal, isCameraOn);
      }
    });
  }

  @Override
  public void onVideoTileAdded(@NonNull VideoTileState tileState) {
    if (tileState.isContent()) {
      handleShareVideoTileAdded(tileState.getAttendeeId());
    } else {
      handleVideoTileAdded(tileState);
    }
  }

  private void handleShareVideoTileAdded(String attendeeId) {
    RosterAttendee attendee = meetingModel().getAttendee(attendeeId);
    boolean isCameraOn = meetingModel().isCameraAttendeeOn(attendeeId);
    sendMeetingUserShareEvent(getReactApplicationContext(), MEETING_ACTIVE_SHARE, attendee, isCameraOn, SHARE_STATUS_START);
  }

  private void handleVideoTileAdded(VideoTileState tile) {
    Log.d(TAG, "onVideoTileAdded tile id: " + tile.getTileId());
    meetingModel().addVideoTile(new VideoCollectionTile(tile));
    sendMeetingUserEvent(getReactApplicationContext(), MEETING_VIDEO_STATUS_CHANGE, meetingModel().getAttendee(tile.getAttendeeId()), true);
  }

  @Override
  public void onVideoTilePaused(@NonNull @NotNull VideoTileState videoTileState) {
    Log.d(TAG, "onVideoTilePaused: " + videoTileState.getTileId() + " state " + videoTileState.getPauseState().name());
  }

  @Override
  public void onVideoTileResumed(@NonNull @NotNull VideoTileState videoTileState) {
    Log.d(TAG, "onVideoTileResumed: " + videoTileState.getTileId());
  }

  @Override
  public void onVideoTileRemoved(@NonNull VideoTileState tileState) {
    if (tileState.isContent()) {
      handleShareVideoTileRemoved(tileState.getAttendeeId());
    } else {
      handleVideoTileRemoved(tileState);
    }
  }

  private void handleShareVideoTileRemoved(String attendeeId) {
    RosterAttendee attendee = meetingModel().getAttendee(attendeeId);
    boolean isCameraOn = meetingModel().isCameraAttendeeOn(attendeeId);
    sendMeetingUserShareEvent(getReactApplicationContext(), MEETING_ACTIVE_SHARE, attendee, isCameraOn, SHARE_STATUS_STOP);
  }

  private void handleVideoTileRemoved(VideoTileState tile) {
    Log.d(TAG, "onVideoTileRemoved tile id: " + tile.getTileId());
    sendMeetingUserEvent(getReactApplicationContext(), MEETING_VIDEO_STATUS_CHANGE, meetingModel().getAttendee(tile.getAttendeeId()), false);
    meetingModel().removeVideoTile(tile.getTileId());
  }

  @Override
  public void onAttendeesMuted(@NonNull AttendeeInfo[] attendeeInfos) {
    Stream.of(attendeeInfos).forEach(it -> {
        RosterAttendee current = meetingModel().getAttendee(it.getAttendeeId());
        RosterAttendee attendee = meetingModel().getCurrentRoster().put(it.getAttendeeId(),
          new RosterAttendee(
            current.getAttendeeId(),
            current.getAttendeeName(),
            true
          ));
        boolean isCameraOn = meetingModel().isCameraAttendeeOn(it.getAttendeeId());
        sendMeetingUserEvent(getReactApplicationContext(), MEETING_AUDIO_STATUS_CHANGE, attendee, isCameraOn);
      }
    );
  }

  @Override
  public void onAttendeesUnmuted(@NonNull AttendeeInfo[] attendeeInfos) {
    Stream.of(attendeeInfos).forEach(it -> {
        RosterAttendee current = meetingModel().getAttendee(it.getAttendeeId());
        RosterAttendee attendee = meetingModel().getCurrentRoster().put(it.getAttendeeId(),
          new RosterAttendee(
            current.getAttendeeId(),
            current.getAttendeeName(),
            true
          ));
        boolean isCameraOn = meetingModel().isCameraAttendeeOn(it.getAttendeeId());
        sendMeetingUserEvent(getReactApplicationContext(), MEETING_AUDIO_STATUS_CHANGE, attendee, isCameraOn);
      }
    );
  }

  @Override
  public void onHostResume() {
    getReactApplicationContext().registerReceiver(moduleConfigReceiver, intentFilter);
  }

  @Override
  public void onHostPause() {
  }

  @Override
  public void onHostDestroy() {
    getReactApplicationContext().unregisterReceiver(moduleConfigReceiver);
    meetingModel().endMeeting();
  }

  @Override
  public void onContentShareStarted() {
  }

  @Override
  public void onContentShareStopped(@NotNull ContentShareStatus contentShareStatus) {

  }
}

File: android/src/main/java/com/reactnativechimesdk/ChimeSdkModule.kt
package com.reactnativechimesdk

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.os.Handler
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.gl.DefaultEglCoreFactory
import com.amazonaws.services.chime.sdk.meetings.session.*
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.amazonaws.services.chime.sdk.meetings.utils.logger.LogLevel
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.google.gson.Gson
import com.reactnativechimesdk.EventEmitter.Companion.KEY_AUDIO_STATUS
import com.reactnativechimesdk.EventEmitter.Companion.KEY_USER_ID
import com.reactnativechimesdk.EventEmitter.Companion.KEY_USER_NAME
import com.reactnativechimesdk.EventEmitter.Companion.MEETING_AUDIO_STATUS_CHANGE
import com.reactnativechimesdk.EventEmitter.Companion.MEETING_USER_JOIN
import com.reactnativechimesdk.EventEmitter.Companion.MEETING_USER_LEFT
import com.reactnativechimesdk.EventEmitter.Companion.sendMeetingStateEvent
import com.reactnativechimesdk.EventEmitter.Companion.sendMeetingUserEvent
import com.reactnativechimesdk.data.JoinMeetingResponse
import com.reactnativechimesdk.data.RosterAttendee
import com.reactnativechimesdk.data.VideoCollectionTile
import com.reactnativechimesdk.itf.SimpleAudioVideoObserver
import com.reactnativechimesdk.itf.SimpleRealTimeObserver
import com.reactnativechimesdk.itf.SimpleVideoTileObserver
import com.reactnativechimesdk.utils.encodeURLParam
import com.reactnativechimesdk.utils.getAttendeeName
import kotlinx.coroutines.*
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

@ReactModule(name = ChimeSdkModule.NAME)
class ChimeSdkModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext),
  LifecycleEventListener,
  SimpleRealTimeObserver,
  SimpleAudioVideoObserver,
  SimpleVideoTileObserver {

  companion object {
    const val TAG = "ChimeSdkModule"
    const val NAME = "ChimeSdk"
    const val WEBRTC_PERMISSION_REQUEST_CODE = 1
    const val MEETING_REGION = "us-east-1"
  }

  private val logger = ConsoleLogger(LogLevel.INFO)
  private val gson = Gson()
  private val uiScope = CoroutineScope(Dispatchers.Main)
  private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO

  private val WEBRTC_PERM = arrayOf(
    Manifest.permission.MODIFY_AUDIO_SETTINGS,
    Manifest.permission.RECORD_AUDIO,
    Manifest.permission.CAMERA
  )

  private val filter: IntentFilter = object : IntentFilter() {
    init {
      addAction("onConfigurationChanged")
      addAction("onRequestPermissionsResult")
    }
  }

  init {
    reactContext.addLifecycleEventListener(this@ChimeSdkModule)
  }

  private lateinit var meetingUrl: String
  private lateinit var meetingId: String
  private lateinit var selfAttendeeName: String

  override fun getName(): String = NAME

  @ReactMethod
  private fun showToast(msg: String) {
    Toast.makeText(
      reactApplicationContext,
      msg,
      Toast.LENGTH_LONG
    ).show()
  }

  @ReactMethod
  fun multiply(a: Int, b: Int, promise: Promise) {
    promise.resolve(a * b)
  }

  @ReactMethod
  fun joinMeeting(map: ReadableMap) {
    meetingUrl = map.getString("meetingUrl").toString()
    meetingId = map.getString("meetingId").toString()
    selfAttendeeName = map.getString("attendeeName").toString()

    if (hasPermissionsAlready()) {
      joinMeeting()
    } else {
      reactContext.currentActivity?.let {
        ActivityCompat.requestPermissions(
          it,
          WEBRTC_PERM,
          WEBRTC_PERMISSION_REQUEST_CODE
        )
      }
    }
  }

  private fun hasPermissionsAlready(): Boolean {
    return WEBRTC_PERM.all {
      reactContext.currentActivity?.let { it1 ->
        ContextCompat.checkSelfPermission(
          it1,
          it
        )
      } == PackageManager.PERMISSION_GRANTED
    }
  }

  private fun joinMeeting() {
    uiScope.launch {
      val meetingResponseJson: String? = requestJoin()
      createSession(meetingResponseJson)
    }
  }

  private suspend fun requestJoin(): String? {
    return withContext(ioDispatcher) {
      val serverUrl =
        URL(
          "${meetingUrl}join?title=${encodeURLParam(meetingId)}" +
            "&name=${encodeURLParam(selfAttendeeName)}" +
            "&region=${encodeURLParam(MEETING_REGION)}"
        )

      try {
        val response = StringBuffer()
        with(serverUrl.openConnection() as HttpURLConnection) {
          requestMethod = "POST"
          doInput = true
          doOutput = true

          BufferedReader(InputStreamReader(inputStream)).use {
            var inputLine = it.readLine()
            while (inputLine != null) {
              response.append(inputLine)
              inputLine = it.readLine()
            }
            it.close()
          }

          if (responseCode == 201) {
            response.toString()
          } else {
            logger.error(TAG, "Unable to join meeting. Response code: $responseCode")
            null
          }
        }
      } catch (exception: Exception) {
        logger.error(TAG, "There was an exception while joining the meeting: $exception")
        null
      }
    }
  }

  private fun createSession(meetingResponseJson: String?) {
    if (meetingResponseJson == null) {
      showToast("Error no meeting response")
    } else {
      val sessionConfig = createSessionConfiguration(meetingResponseJson)
      val meetingSession = sessionConfig?.let {
        DefaultMeetingSession(
          it,
          logger,
          reactApplicationContext,
          // Note if the following isn't provided app will (as expected) crash if we use custom video source
          // since an EglCoreFactory will be internal created and will be using a different shared EGLContext.
          // However the internal default capture would work fine, since it is initialized using
          // that internally created default EglCoreFactory, and can be smoke tested by removing this
          // argument and toggling use of custom video source before starting video
          DefaultEglCoreFactory()
        )
      }

      if (meetingSession == null) {
        showToast("Error null meeting session")
        return
      }

      MeetingSessionModel.setMeetingSession(meetingSession)
      MeetingSessionModel.audioVideo.apply {
        addRealtimeObserver(this@ChimeSdkModule)
        addAudioVideoObserver(this@ChimeSdkModule)
        addVideoTileObserver(this@ChimeSdkModule)
        start()
        startLocalVideo()
        startRemoteVideo()
      }
    }
  }

  private fun createSessionConfiguration(response: String?): MeetingSessionConfiguration? {
    if (response.isNullOrBlank()) return null

    return try {
      val joinMeetingResponse = gson.fromJson(response, JoinMeetingResponse::class.java)
      MeetingSessionConfiguration(
        CreateMeetingResponse(joinMeetingResponse.joinInfo.meetingResponse.meeting),
        CreateAttendeeResponse(joinMeetingResponse.joinInfo.attendeeResponse.attendee),
        ::urlRewriter
      )
    } catch (exception: Exception) {
      logger.error(
        TAG,
        "Error creating session configuration: ${exception.localizedMessage}"
      )
      null
    }
  }

  private fun urlRewriter(url: String): String {
    // You can change urls by url.replace("example.com", "my.example.com")
    return url
  }

  @ReactMethod
  fun leaveCurrentMeeting() {
    sendMeetingStateEvent(reactContext, "meeting_end")
    Handler().postDelayed({
      sendMeetingStateEvent(reactContext, "idle")
    }, 1500);
    endMeeting()
  }

  private fun endMeeting() {
    MeetingSessionModel.audioVideo.apply {
      stopLocalVideo()
      stopRemoteVideo()
      stop()
    }
    MeetingSessionModel.currentRoster.clear()
  }

  @ReactMethod
  fun getParticipants(callback: Callback) {
    val array: WritableArray = WritableNativeArray()
    MeetingSessionModel.currentRoster.values.forEach {
      val map: WritableMap = WritableNativeMap()
      map.putString(KEY_USER_ID, it.attendeeId)
      map.putString(KEY_USER_NAME, it.attendeeName)
      // map.putBoolean(KEY_VIDEO_STATUS, it.isCameraOn)
      map.putBoolean(KEY_AUDIO_STATUS, it.isMuted)
      array.pushMap(map)
    }
    callback.invoke(array)
  }

  @ReactMethod
  fun getUserInfo(attendeeId: String, callback: Callback) {
    MeetingSessionModel.currentRoster.values.find { it.attendeeId == attendeeId }?.let {
      val map: WritableMap = WritableNativeMap()
      map.putString(KEY_USER_ID, it.attendeeId)
      map.putString(KEY_USER_NAME, it.attendeeName)
      // map.putBoolean(KEY_VIDEO_STATUS, it.isCameraOn)
      map.putBoolean(KEY_AUDIO_STATUS, it.isMuted)
      callback.invoke(map)
    }
  }

  @ReactMethod
  fun onMyAudio() {
    MeetingSessionModel.audioVideo.realtimeLocalMute()
  }

  @ReactMethod
  fun offMyAudio() {
    MeetingSessionModel.audioVideo.realtimeLocalUnmute()
  }

  @ReactMethod
  fun onOffMyVideo() {
    if (MeetingSessionModel.isCameraOn) {
      MeetingSessionModel.audioVideo.stopLocalVideo()
    } else {
      MeetingSessionModel.audioVideo.startLocalVideo()
    }
  }

  @ReactMethod
  fun switchMyCamera() {
  }

  override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {
    super.onAttendeesMuted(attendeeInfo)
    attendeeInfo.forEach { attendee ->
      MeetingSessionModel.currentRoster[attendee.attendeeId]?.let {
        MeetingSessionModel.currentRoster[it.attendeeId] =
          RosterAttendee(
            it.attendeeId,
            it.attendeeName,
            it.volumeLevel,
            it.signalStrength,
            it.isActiveSpeaker,
            true
          )
        sendMeetingUserEvent(reactContext, MEETING_AUDIO_STATUS_CHANGE, it)
      }
    }
  }

  override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {
    super.onAttendeesUnmuted(attendeeInfo)
    attendeeInfo.forEach { attendee ->
      MeetingSessionModel.currentRoster[attendee.attendeeId]?.let {
        MeetingSessionModel.currentRoster[it.attendeeId] =
          RosterAttendee(
            it.attendeeId,
            it.attendeeName,
            it.volumeLevel,
            it.signalStrength,
            it.isActiveSpeaker,
            false
          )
        sendMeetingUserEvent(reactContext, MEETING_AUDIO_STATUS_CHANGE, it)
      }
    }
  }

  override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
    super.onAttendeesJoined(attendeeInfo)
    attendeeInfo.forEach {
      MeetingSessionModel.currentRoster.getOrPut(
        it.attendeeId,
        { RosterAttendee(it.attendeeId, getAttendeeName(it.attendeeId, it.externalUserId)) })

      if (MeetingSessionModel.localVideoTileState?.videoTileState?.attendeeId == it.attendeeId) {
        sendMeetingStateEvent(reactContext, "meeting_ready")
      } else {
        sendMeetingUserEvent(reactContext, MEETING_USER_JOIN, MeetingSessionModel.currentRoster[it.attendeeId]!!)
      }
    }
  }

  override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
    super.onAttendeesLeft(attendeeInfo)
    attendeeInfo.forEach {
      MeetingSessionModel.currentRoster[it.attendeeId]?.let { attendee ->
        sendMeetingUserEvent(reactContext, MEETING_USER_LEFT, attendee)
      }
      MeetingSessionModel.currentRoster.remove(
        it.attendeeId
      )
    }
  }

  override fun onAudioSessionStarted(reconnecting: Boolean) {
    super.onAudioSessionStarted(reconnecting)
    val success = MeetingSessionModel.audioVideo.realtimeSetVoiceFocusEnabled(true)
    if (success) {
      Log.d(TAG, "Voice Focus enabled")
    } else {
      Log.d(TAG, "Failed to enable Voice Focus")
    }
  }

  override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
    super.onAudioSessionStopped(sessionStatus)
  }

  override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {
    super.onVideoSessionStarted(sessionStatus)
  }

  override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {
    super.onVideoSessionStopped(sessionStatus)
  }

  override fun onVideoTileAdded(tileState: VideoTileState) {
    super.onVideoTileAdded(tileState)
    uiScope.launch {
      val videoCollectionTile = createVideoCollectionTile(tileState)
      if (tileState.isLocalTile) {
        MeetingSessionModel.localAttendeeId = tileState.attendeeId
        MeetingSessionModel.isCameraOn = true
        MeetingSessionModel.localVideoTileState = videoCollectionTile
      } else {
        MeetingSessionModel.addRemoteVideoTile(videoCollectionTile)
      }
    }
  }

  private fun createVideoCollectionTile(tileState: VideoTileState): VideoCollectionTile {
    val attendeeId = tileState.attendeeId
    val attendeeName = MeetingSessionModel.currentRoster[attendeeId]?.attendeeName ?: ""
    return VideoCollectionTile(attendeeName, tileState)
  }

  override fun onVideoTileRemoved(tileState: VideoTileState) {
    super.onVideoTileRemoved(tileState)
    uiScope.launch {
      if (tileState.isLocalTile) {
        MeetingSessionModel.localVideoTileState = null
        MeetingSessionModel.isCameraOn = false
      } else {
        MeetingSessionModel.removeRemoteVideoTile(tileState.tileId)
      }
    }
  }

  override fun onHostResume() {
    reactContext.registerReceiver(moduleConfigReceiver, filter)
  }

  override fun onHostPause() {
    // Ignore
  }

  override fun onHostDestroy() {
    reactContext.unregisterReceiver(moduleConfigReceiver)
  }

  private val moduleConfigReceiver: BroadcastReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
      if ("onConfigurationChanged" == intent.action) {
        val configuration = intent.getParcelableExtra<Configuration>("newConfig")
        if (configuration != null) {
          // refreshRotation()
        }
      } else if ("onRequestPermissionsResult" == intent.action) {
        val permissions = intent.getStringArrayExtra("permissions")
        val grantResults = intent.getIntArrayExtra("grantResults")
        if (permissions == null || grantResults == null) {
          return
        }
        joinMeeting()
      }
    }
  }
}
